lib/cloud/gcp/imds.go

```diff
@@ -43,6 +50,149 @@ func (rt contextRoundTripper) RoundTrip(req *http.Request) (*http.Response, erro
 	return resp, trace.Wrap(err)
 }
 
+type metadataGetter func(ctx context.Context, path string) (string, error)
+
+// InstanceMetadataClient is a client for GCP instance metadata.
+type InstanceMetadataClient struct {
+	instancesClient InstancesClient
+	getMetadata     metadataGetter
+
+	labelPermissionErrorOnce sync.Once
+	tagPermissionErrorOnce   sync.Once
+}
+
+// NewInstanceMetadataClient creates a new instance metadata client.
+func NewInstanceMetadataClient(ctx context.Context) (*InstanceMetadataClient, error) {
+	instancesClient, err := NewInstancesClient(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &InstanceMetadataClient{
+		instancesClient: instancesClient,
+		getMetadata:     getMetadata,
+	}, nil
+}
+
+// IsAvailable checks if instance metadata is available.
+func (client *InstanceMetadataClient) IsAvailable(ctx context.Context) bool {
+	instanceData, err := client.getMetadata(ctx, "instance")
+	return err == nil && instanceData != ""
+}
+
+// GetTags gets all of the GCP instance's labels (note: these are separate from
+// its tags, which we do not use).
+func (client *InstanceMetadataClient) GetTags(ctx context.Context) (map[string]string, error) {
+	// Get a bunch of info from instance metadata.
+	projectID, err := client.GetProjectID(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	zone, err := client.GetZone(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	name, err := client.GetName(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	idStr, err := client.GetID(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	id, err := strconv.ParseUint(idStr, 10, 64)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	req := &InstanceRequest{
+		ProjectID:       projectID,
+		Zone:            zone,
+		Name:            name,
+		ID:              id,
+		WithoutHostKeys: true,
+	}
+	// Get labels.
+	var gcpLabels map[string]string
+	inst, err := client.instancesClient.GetInstance(ctx, req)
+	if err == nil {
+		gcpLabels = inst.Labels
+	} else if trace.IsAccessDenied(err) {
```

Comment:
I don't see the place where we convert the error from gcp error to trace kind of error so this assertion won't work

Reply:
We convert it [here](https://github.com/gravitational/teleport/blob/631f3ccef6afe69a37e6784ae587c93231f7ed26/lib/cloud/gcp/vm.go#L321-L323). But there are a few places where that wrapper is missing, so I've added it in.